### ISSUE TITLE
Python3 compatibility issue: natsort.natsort(['1-1', 'aa1-2'])

### ISSUE THREAD

**Issue by dmr** [xolox/python-naturalsort#2]
Hi,

I tried to use https://github.com/xolox/python-rotate-backups with python3 and it wasn't working because of an issue/specific behavior in natsort:

With python2, `natsort.natsort(['1-1', 'aa1-2'])`returns a result.
With python3 however, not: A TypeError is raised:

```
.../natsort/__init__.py in natsort(l, key, reverse)
     17     """
     18     return sorted(l, key=lambda v: natsort_key(key and key(v) or v),
---> 19                   reverse=reverse)
     20 
     21 def natsort_key(s):

TypeError: unorderable types: str() < int()
```

Do you know how to fix this?

**Comment by dmr** [comment]
The issue can be "solved" with a string cast in natsort_key:

```
def natsort_key(s):
    """
    Turn a string into a list of substrings and numbers.
    """
    return [str(coerce(c)) for c in integer_pattern.split(s) if c != '']
```

works with python3.

So I can monkey-patch it now:

```
natsort.natsort_key = lambda s: [
    str(int(s) if s.isdigit() else s)
    for s in natsort.integer_pattern.split(s) if s != ''
]
```

But it doesn't feel right ^^.

**Comment by xolox** [comment]
Hi Daniel,

I noticed this issue as well recently, wrote it down as a "note to self" but hadn't created a public issue for it yet. Thanks for reporting this! This change in semantics of Python 2 versus Python 3 actually surprised me because it breaks certain use cases of the well known Python [decorate-sort-undecorate](https://wiki.python.org/moin/HowTo/Sorting#The_Old_Way_Using_Decorate-Sort-Undecorate) idiom.

FYI: Your fix changes the semantics so it no longer uses natural order sorting. Whether that's a serious problem for you is another topic, basically it depends on the filenames you use and whether they contain zero padded numbers.

I'll see how I can fix this and report back here soon.

**Comment by dmr** [comment]
Thank you for the quick reply. I was wondering if my hack was working, sorry to hear that it is not.
I ended up switching to another rotate script instead, so it's not that important for me to have natsort work with python3 anymore.

Regarding python3 I agree: Lots of weird breaking changes. I had most of the problems in my projects with unicode an/or Bytestrings vs Strings but I came to conclusion that python3 really is better and switched completely dropping python2 support eventually.

**Comment by xolox** [comment]
Hi Daniel,

The problem you reported should now be fixed (I broke the tests in 1da3dd8 to reproduce the problem and the fixes in cd12088 which didn't touch the test suite made the tests work again) so I'm closing this issue. Thanks again for reporting the issue!

I've also restored Python 3 support for the rotate-backups package in xolox/python-rotate-backups@f46d3d6. Since there was no separate issue for that I thought I would mention it here :-).
